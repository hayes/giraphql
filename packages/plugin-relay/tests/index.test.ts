import { gql } from 'apollo-server';
import { execute, lexicographicSortSchema, printSchema } from 'graphql';
import schema from './examples/relay/schema';

describe('relay example schema', () => {
  test('generates expected schema', () => {
    expect(printSchema(lexicographicSortSchema(schema))).toMatchSnapshot();
  });

  describe('queries', () => {
    test('query offset based connections', async () => {
      const query = gql`
        {
          numbers(first: 3) {
            pageInfo {
              hasNextPage
              hasPreviousPage
              startCursor
              endCursor
            }
            edges {
              cursor
              node {
                number
              }
            }
          }
          firstAfterAtEnd: numbers(first: 3, after: "T2Zmc2V0Q29ubmVjdGlvbjoxOTc=") {
            pageInfo {
              hasNextPage
              hasPreviousPage
              startCursor
              endCursor
            }
            edges {
              cursor
              node {
                number
              }
            }
          }
          firstAfter: numbers(first: 3, after: "T2Zmc2V0Q29ubmVjdGlvbjox") {
            pageInfo {
              hasNextPage
              hasPreviousPage
              startCursor
              endCursor
            }
            edges {
              cursor
              node {
                number
              }
            }
          }
          lastAtBegining: numbers(before: "T2Zmc2V0Q29ubmVjdGlvbjox") {
            pageInfo {
              hasNextPage
              hasPreviousPage
              startCursor
              endCursor
            }
            edges {
              cursor
              node {
                number
              }
            }
          }
          lastBefore: numbers(last: 3, before: "T2Zmc2V0Q29ubmVjdGlvbjoxMA==") {
            pageInfo {
              hasNextPage
              hasPreviousPage
              startCursor
              endCursor
            }
            edges {
              cursor
              node {
                number
              }
            }
          }
          beforeAndAfter: numbers(
            before: "T2Zmc2V0Q29ubmVjdGlvbjoxMA=="
            after: "T2Zmc2V0Q29ubmVjdGlvbjo3"
          ) {
            pageInfo {
              hasNextPage
              hasPreviousPage
              startCursor
              endCursor
            }
            edges {
              cursor
              node {
                number
              }
            }
          }
          firstAndLast: numbers(first: 5, last: 2, before: "T2Zmc2V0Q29ubmVjdGlvbjoxMA==") {
            pageInfo {
              hasNextPage
              hasPreviousPage
              startCursor
              endCursor
            }
            edges {
              cursor
              node {
                number
              }
            }
          }
        }
      `;

      const result = await execute({
        schema,
        document: query,
        contextValue: {},
      });

      expect(result).toMatchSnapshot();
    });

    test('query array based connections', async () => {
      const query = gql`
        {
          batchNumbers(first: 3) {
            pageInfo {
              hasNextPage
              hasPreviousPage
              startCursor
              endCursor
            }
            edges {
              cursor
              node {
                number
              }
            }
          }
          firstAfterAtEnd: batchNumbers(first: 3, after: "T2Zmc2V0Q29ubmVjdGlvbjoxOTc=") {
            pageInfo {
              hasNextPage
              hasPreviousPage
              startCursor
              endCursor
            }
            edges {
              cursor
              node {
                number
              }
            }
          }
          firstAfter: batchNumbers(first: 3, after: "T2Zmc2V0Q29ubmVjdGlvbjox") {
            pageInfo {
              hasNextPage
              hasPreviousPage
              startCursor
              endCursor
            }
            edges {
              cursor
              node {
                number
              }
            }
          }
          lastAtBegining: batchNumbers(before: "T2Zmc2V0Q29ubmVjdGlvbjox") {
            pageInfo {
              hasNextPage
              hasPreviousPage
              startCursor
              endCursor
            }
            edges {
              cursor
              node {
                number
              }
            }
          }
          lastBefore: batchNumbers(last: 3, before: "T2Zmc2V0Q29ubmVjdGlvbjoxMA==") {
            pageInfo {
              hasNextPage
              hasPreviousPage
              startCursor
              endCursor
            }
            edges {
              cursor
              node {
                number
              }
            }
          }
          beforeAndAfter: batchNumbers(
            before: "T2Zmc2V0Q29ubmVjdGlvbjoxMA=="
            after: "T2Zmc2V0Q29ubmVjdGlvbjo3"
          ) {
            pageInfo {
              hasNextPage
              hasPreviousPage
              startCursor
              endCursor
            }
            edges {
              cursor
              node {
                number
              }
            }
          }
          firstAndLast: batchNumbers(first: 5, last: 2, before: "T2Zmc2V0Q29ubmVjdGlvbjoxMA==") {
            pageInfo {
              hasNextPage
              hasPreviousPage
              startCursor
              endCursor
            }
            edges {
              cursor
              node {
                number
              }
            }
          }
        }
      `;

      const result = await execute({
        schema,
        document: query,
        contextValue: {},
      });

      expect(result).toMatchSnapshot();
    });

    test('query node and nodes on query type', async () => {
      const query = gql`
        {
          extraNode {
            __typename
            ... on Number {
              number
            }
          }
          moreNodes {
            __typename
            ... on BatchNumber {
              batchNumber: number
            }
            ... on Number {
              number
            }
          }
          nodes(ids: ["TnVtYmVyOjQ=", "TnVtYmVyOjI=", "TnVtYmVyOjI=", "QmF0Y2hOdW1iZXI6MA=="]) {
            ... on BatchNumber {
              batchNumber: number
            }
            ... on Number {
              number
            }
          }

          batchNode: node(id: "QmF0Y2hOdW1iZXI6MA==") {
            ... on BatchNumber {
              number
            }
          }
          node(id: "TnVtYmVyOjI=") {
            id
            ... on Number {
              number
            }
          }

          node2: node(id: "TnVtYmVyOjI=") {
            ... on Number {
              number
            }
          }
        }
      `;

      const result = await execute({
        schema,
        document: query,
        contextValue: {},
      });

      expect(result).toMatchSnapshot();
    });

    test('globalID inputs', async () => {
      const query = gql`
        {
          inputGlobalID(
            id: "YWJjOjEyMw=="
            normalId: 123
            inputObj: {
              id: "YWJjOjEyMw=="
              idList: ["YWJjOjEyMw=="]
              circular: { id: "YWJjOjEyMw==", idList: ["YWJjOjEyMw=="] }
            }
          )
        }
      `;

      const result = await execute({
        schema,
        document: query,
        contextValue: {},
      });

      expect(JSON.parse(result.data!.inputGlobalID)).toMatchInlineSnapshot(`
        Object {
          "id": Object {
            "id": "123",
            "typename": "abc",
          },
          "inputObj": Object {
            "circular": Object {
              "id": Object {
                "id": "123",
                "typename": "abc",
              },
              "idList": Array [
                Object {
                  "id": "123",
                  "typename": "abc",
                },
              ],
            },
            "id": Object {
              "id": "123",
              "typename": "abc",
            },
            "idList": Array [
              Object {
                "id": "123",
                "typename": "abc",
              },
            ],
          },
          "normal": "123",
        }
      `);
    });
  });
});
