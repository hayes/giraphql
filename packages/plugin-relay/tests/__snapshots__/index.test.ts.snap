// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`relay example schema generates expected schema 1`] = `
"type Answer {
  count: Int!
  id: ID!
  value: String!
}

type BatchNumber implements Node {
  id: ID!
  number: Int!
}

input GlobalIDInput {
  circular: GlobalIDInput
  circularWithoutGlobalIds: NoGlobalIDInput
  id: ID!
  idList: [ID!]!
}

type Mutation {
  answerPoll(answer: Int!, id: ID!): Poll!
  createPoll(answers: [String!]!, question: String!): Poll!
}

input NoGlobalIDInput {
  circular: NoGlobalIDInput
  id: ID
}

interface Node {
  id: ID!
}

type Number implements Node {
  id: ID!
  number: Int!
}

type PageInfo {
  endCursor: String
  hasNextPage: Boolean!
  hasPreviousPage: Boolean!
  startCursor: String
}

type Poll implements Node {
  answers: [Answer!]!
  answersConnection(after: ID, before: ID, first: Int, last: Int): PollAnswersConnection!
  answersUsingOffset(after: ID, before: ID, first: Int, last: Int): PollAnswersUsingOffsetConnection!
  answersWithoutHelpers(after: ID, before: ID, first: Int, last: Int): PollAnswersCon!
  id: ID!
  question: String!
  updatedAt: String!
}

type PollAnswersCon {
  edges: [PollAnswersConEdge]!
  pageInfo: PageInfo!
}

type PollAnswersConEdge {
  cursor: String!
  node: Answer!
}

type PollAnswersConnection {
  edges: [PollAnswersConnectionEdge]!
  pageInfo: PageInfo!
}

type PollAnswersConnectionEdge {
  cursor: String!
  node: Answer!
}

type PollAnswersUsingOffsetConnection {
  edges: [PollAnswersUsingOffsetConnectionEdge]!
  pageInfo: PageInfo!
}

type PollAnswersUsingOffsetConnectionEdge {
  cursor: String!
  node: Answer!
}

type Query {
  batchNumbers(after: ID, before: ID, first: Int, last: Int): QueryBatchNumbersConnection!
  extraNode: Node
  inputGlobalID(id: ID!, inputObj: GlobalIDInput!, normalId: ID!): String!
  moreNodes: [Node]!
  node(id: ID!): Node
  nodes(ids: [ID!]!): [Node]!
  numbers(after: ID, before: ID, first: Int, last: Int): QueryNumbersConnection!
  poll(id: Int!): Poll
  pollIds: [ID!]!
  polls: [Poll!]!
  pollsConnection(after: ID, before: ID, first: Int, last: Int): QueryPollsConnection!
}

type QueryBatchNumbersConnection {
  edges: [QueryBatchNumbersConnectionEdge]!
  pageInfo: PageInfo!
}

type QueryBatchNumbersConnectionEdge {
  cursor: String!
  node: BatchNumber!
}

type QueryNumbersConnection {
  edges: [QueryNumbersConnectionEdge]!
  pageInfo: PageInfo!
}

type QueryNumbersConnectionEdge {
  cursor: String!
  node: Number!
}

type QueryPollsConnection {
  edges: [QueryPollsConnectionEdge]!
  extraConnectionField: String!
  pageInfo: PageInfo!
}

type QueryPollsConnectionEdge {
  cursor: String!
  extraEdgeField: Int!
  node: Poll!
}
"
`;

exports[`relay example schema queries query array based connections 1`] = `
Object {
  "data": Object {
    "batchNumbers": Object {
      "edges": Array [
        Object {
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjow",
          "node": Object {
            "number": 0,
          },
        },
        Object {
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjox",
          "node": Object {
            "number": 1,
          },
        },
        Object {
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjoy",
          "node": Object {
            "number": 2,
          },
        },
      ],
      "pageInfo": Object {
        "endCursor": "T2Zmc2V0Q29ubmVjdGlvbjoy",
        "hasNextPage": true,
        "hasPreviousPage": false,
        "startCursor": "T2Zmc2V0Q29ubmVjdGlvbjow",
      },
    },
    "beforeAndAfter": Object {
      "edges": Array [
        Object {
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjo4",
          "node": Object {
            "number": 8,
          },
        },
        Object {
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjo5",
          "node": Object {
            "number": 9,
          },
        },
      ],
      "pageInfo": Object {
        "endCursor": "T2Zmc2V0Q29ubmVjdGlvbjo5",
        "hasNextPage": true,
        "hasPreviousPage": true,
        "startCursor": "T2Zmc2V0Q29ubmVjdGlvbjo4",
      },
    },
    "firstAfter": Object {
      "edges": Array [
        Object {
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjoy",
          "node": Object {
            "number": 2,
          },
        },
        Object {
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjoz",
          "node": Object {
            "number": 3,
          },
        },
        Object {
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjo0",
          "node": Object {
            "number": 4,
          },
        },
      ],
      "pageInfo": Object {
        "endCursor": "T2Zmc2V0Q29ubmVjdGlvbjo0",
        "hasNextPage": true,
        "hasPreviousPage": true,
        "startCursor": "T2Zmc2V0Q29ubmVjdGlvbjoy",
      },
    },
    "firstAfterAtEnd": Object {
      "edges": Array [
        Object {
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjoxOTg=",
          "node": Object {
            "number": 198,
          },
        },
        Object {
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjoxOTk=",
          "node": Object {
            "number": 199,
          },
        },
      ],
      "pageInfo": Object {
        "endCursor": "T2Zmc2V0Q29ubmVjdGlvbjoxOTk=",
        "hasNextPage": false,
        "hasPreviousPage": true,
        "startCursor": "T2Zmc2V0Q29ubmVjdGlvbjoxOTg=",
      },
    },
    "firstAndLast": Object {
      "edges": Array [
        Object {
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjoz",
          "node": Object {
            "number": 3,
          },
        },
        Object {
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjo0",
          "node": Object {
            "number": 4,
          },
        },
      ],
      "pageInfo": Object {
        "endCursor": "T2Zmc2V0Q29ubmVjdGlvbjo0",
        "hasNextPage": true,
        "hasPreviousPage": true,
        "startCursor": "T2Zmc2V0Q29ubmVjdGlvbjoz",
      },
    },
    "lastAtBegining": Object {
      "edges": Array [
        Object {
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjow",
          "node": Object {
            "number": 0,
          },
        },
      ],
      "pageInfo": Object {
        "endCursor": "T2Zmc2V0Q29ubmVjdGlvbjow",
        "hasNextPage": true,
        "hasPreviousPage": false,
        "startCursor": "T2Zmc2V0Q29ubmVjdGlvbjow",
      },
    },
    "lastBefore": Object {
      "edges": Array [
        Object {
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjo3",
          "node": Object {
            "number": 7,
          },
        },
        Object {
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjo4",
          "node": Object {
            "number": 8,
          },
        },
        Object {
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjo5",
          "node": Object {
            "number": 9,
          },
        },
      ],
      "pageInfo": Object {
        "endCursor": "T2Zmc2V0Q29ubmVjdGlvbjo5",
        "hasNextPage": false,
        "hasPreviousPage": true,
        "startCursor": "T2Zmc2V0Q29ubmVjdGlvbjo3",
      },
    },
  },
}
`;

exports[`relay example schema queries query node and nodes on query type 1`] = `
Object {
  "data": Object {
    "batchNode": Object {
      "number": 0,
    },
    "extraNode": Object {
      "__typename": "Number",
      "number": 2,
    },
    "moreNodes": Array [
      Object {
        "__typename": "Number",
        "number": 2,
      },
      Object {
        "__typename": "Number",
        "number": 10,
      },
    ],
    "node": Object {
      "id": "TnVtYmVyOjI=",
      "number": 2,
    },
    "node2": Object {
      "number": 2,
    },
    "nodes": Array [
      Object {
        "number": 4,
      },
      Object {
        "number": 2,
      },
      Object {
        "number": 2,
      },
      Object {
        "batchNumber": 0,
      },
    ],
  },
}
`;

exports[`relay example schema queries query offset based connections 1`] = `
Object {
  "data": Object {
    "beforeAndAfter": Object {
      "edges": Array [
        Object {
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjo4",
          "node": Object {
            "number": 8,
          },
        },
        Object {
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjo5",
          "node": Object {
            "number": 9,
          },
        },
      ],
      "pageInfo": Object {
        "endCursor": "T2Zmc2V0Q29ubmVjdGlvbjo5",
        "hasNextPage": true,
        "hasPreviousPage": true,
        "startCursor": "T2Zmc2V0Q29ubmVjdGlvbjo4",
      },
    },
    "firstAfter": Object {
      "edges": Array [
        Object {
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjoy",
          "node": Object {
            "number": 2,
          },
        },
        Object {
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjoz",
          "node": Object {
            "number": 3,
          },
        },
        Object {
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjo0",
          "node": Object {
            "number": 4,
          },
        },
      ],
      "pageInfo": Object {
        "endCursor": "T2Zmc2V0Q29ubmVjdGlvbjo0",
        "hasNextPage": true,
        "hasPreviousPage": true,
        "startCursor": "T2Zmc2V0Q29ubmVjdGlvbjoy",
      },
    },
    "firstAfterAtEnd": Object {
      "edges": Array [
        Object {
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjoxOTg=",
          "node": Object {
            "number": 198,
          },
        },
        Object {
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjoxOTk=",
          "node": Object {
            "number": 199,
          },
        },
      ],
      "pageInfo": Object {
        "endCursor": "T2Zmc2V0Q29ubmVjdGlvbjoxOTk=",
        "hasNextPage": false,
        "hasPreviousPage": true,
        "startCursor": "T2Zmc2V0Q29ubmVjdGlvbjoxOTg=",
      },
    },
    "firstAndLast": Object {
      "edges": Array [
        Object {
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjoz",
          "node": Object {
            "number": 3,
          },
        },
        Object {
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjo0",
          "node": Object {
            "number": 4,
          },
        },
      ],
      "pageInfo": Object {
        "endCursor": "T2Zmc2V0Q29ubmVjdGlvbjo0",
        "hasNextPage": true,
        "hasPreviousPage": true,
        "startCursor": "T2Zmc2V0Q29ubmVjdGlvbjoz",
      },
    },
    "lastAtBegining": Object {
      "edges": Array [
        Object {
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjow",
          "node": Object {
            "number": 0,
          },
        },
      ],
      "pageInfo": Object {
        "endCursor": "T2Zmc2V0Q29ubmVjdGlvbjow",
        "hasNextPage": true,
        "hasPreviousPage": false,
        "startCursor": "T2Zmc2V0Q29ubmVjdGlvbjow",
      },
    },
    "lastBefore": Object {
      "edges": Array [
        Object {
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjo3",
          "node": Object {
            "number": 7,
          },
        },
        Object {
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjo4",
          "node": Object {
            "number": 8,
          },
        },
        Object {
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjo5",
          "node": Object {
            "number": 9,
          },
        },
      ],
      "pageInfo": Object {
        "endCursor": "T2Zmc2V0Q29ubmVjdGlvbjo5",
        "hasNextPage": false,
        "hasPreviousPage": true,
        "startCursor": "T2Zmc2V0Q29ubmVjdGlvbjo3",
      },
    },
    "numbers": Object {
      "edges": Array [
        Object {
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjow",
          "node": Object {
            "number": 0,
          },
        },
        Object {
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjox",
          "node": Object {
            "number": 1,
          },
        },
        Object {
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjoy",
          "node": Object {
            "number": 2,
          },
        },
      ],
      "pageInfo": Object {
        "endCursor": "T2Zmc2V0Q29ubmVjdGlvbjoy",
        "hasNextPage": true,
        "hasPreviousPage": false,
        "startCursor": "T2Zmc2V0Q29ubmVjdGlvbjow",
      },
    },
  },
}
`;
